fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -609,10 +608,11 @@ private Tuple recordVersionKey(@Nonnull Tuple primaryKey) {
                 return;
             }
             final Map<RecordType, Collection<IndexMaintainer>> maintainers = getSyntheticMaintainers(plan.getSyntheticRecordTypes());
-            final Map<Tuple, FDBSyntheticRecord> oldRecords = new HashMap<>();
+            // Any synthetic records are processed one at a time, since there is no requirement that index maintainers are thread-safe.
+            final Map<Tuple, FDBSyntheticRecord> oldRecords = new ConcurrentHashMap<>();
             CompletableFuture<Void> future = plan.execute(this, oldRecord).forEach(syntheticRecord -> oldRecords.put(syntheticRecord.getPrimaryKey(), syntheticRecord));
             @Nonnull final FDBStoredRecord<M> theNewRecord = newRecord; // @SpotBugsSuppressWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE", justification = "https://github.com/spotbugs/spotbugs/issues/552")
-            future = future.thenCompose(v -> plan.execute(this, theNewRecord).forEachAsync(syntheticRecord -> runSyntheticMaintainers(maintainers, oldRecords.remove(syntheticRecord.getPrimaryKey()), syntheticRecord), pipelineSize));
+            future = future.thenCompose(v -> plan.execute(this, theNewRecord).forEach(syntheticRecord -> runSyntheticMaintainers(maintainers, oldRecords.remove(syntheticRecord.getPrimaryKey()), syntheticRecord)));
```

Comment:
Won't changing to these to `forEach` actually run everything in parallel (in that it will fire off all of the futures in serial but wait on none of them)?

Reply:
Yes, that sounds right.